subprojects/core/src/main/java/org/gradle/process/internal/DefaultJavaExecSpec.java

```diff
@@ -90,60 +81,42 @@ public void copyTo(JavaExecSpec targetSpec) {
 
     @Override
     public Provider<List<String>> getCommandLine() {
-        return getExecutable().map(executable -> argumentsSpec.getCommandLine());
+        return getExecutable().zip(getAllJvmArgs(), (SerializableLambdas.SerializableBiFunction<String, List<String>, List<String>>) (executable, allJvmArgs) -> {
+            List<String> allArgs = ExecHandleCommandLineCombiner.getAllArgs(allJvmArgs, getArgs().get(), getArgumentProviders().get());
+            return ExecHandleCommandLineCombiner.getCommandLine(executable, allArgs);
+        });
     }
 
     @Override
     public JavaExecSpec args(Object... args) {
-        argumentsSpec.args(args);
+        DefaultExecSpec.collectArgs(getArgs(), args);
```

Comment:
💅 This would look nicer if this `static` method was on a class independent of `DefaultExecSpec` to avoid the dependency between the two classes.

Reply:
Makes sense, moved to `ExecHandleCommandLineCombiner` where we have similar methods

Reply:
Makes sense, moved to `ExecHandleCommandLineCombiner` where we already have such methods